### PR TITLE
[Merged by Bors] - feat: nlinarith

### PR DESCRIPTION
**Pull request by kim-em**

### CODE REVIEW

File: Mathlib/Tactic/Linarith/Frontend.lean [hunk 1]
```diff
@@ -244,7 +244,7 @@ if it does not succeed at doing this.
 partial def linarith (only_on : Bool) (hyps : List Expr) (cfg : LinarithConfig := {})
     (g : MVarId) : MetaM Unit := do
   -- if the target is an equality, we run `linarith` twice, to prove ≤ and ≥.
-  if (← g.getType).isEq then do
+  if (← instantiateMVars (← g.getType)).isEq then do
```
Comment (digama0) [inline Mathlib/Tactic/Linarith/Frontend.lean:247]:
```suggestion
  if (← g.getType').isEq then do
```

File: Mathlib/Tactic/Linarith/Frontend.lean [hunk 2]
```diff
@@ -373,28 +389,37 @@ elab_rules : tactic
...
+    let cfg :=
+    { cfg with
+      preprocessors := some (cfg.preprocessors.getD defaultPreprocessors ++
+        [(nlinarithExtras : GlobalBranchingPreprocessor)]) }
+    liftMetaFinishingTactic <|
+      Linarith.linarith o.isSome
+        (← ((args.map (TSepArray.getElems)).getD {}).mapM (elabTerm ·.raw none)).toList
+        cfg
 
+-- FIXME We have to repeat all that just to handle the `!`?
```
Comment (digama0) [inline Mathlib/Tactic/Linarith/Frontend.lean:406]: The pattern used in other tactics is to have the main tactic take a `$[!]?`, and then have the bang-tactic be a macro that expands as `nlinarith! $args` -> `nlinarith ! $args`.

File: Mathlib/Tactic/Linarith/Preprocessing.lean [hunk 3]
```diff
@@ -301,57 +301,85 @@ section cancelDenoms
...
+That is, for each `e ∈ l`, it will run `f e e` and then `f e e'`
+for each `e'` that appears after `e` in `l`.
 
--- * For every term `t` such that `t^2` or `t*t` appears in the input, adds a proof of `t^2 ≥ 0`
---   or `t*t ≥ 0`.
--- * For every pair of comparisons `t1 R1 0` and `t2 R2 0`, adds a proof of `t1*t2 R 0`.
+Example: suppose `l = [1, 2, 3]`. `mapUpperTriangleM f l` will produce the list
+`[f 1 1, f 1 2, f 1 3, f 2 2, f 2 3, f 3 3]`.
+-/
+def _root_.List.mapUpperTriangleM {m} [Monad m] {α β : Type u} (f : α → α → m β) :
```
Comment (digama0) [inline Mathlib/Tactic/Linarith/Preprocessing.lean:332]: This is `mapDiagM`

**Review by digama0 — APPROVED** [review]

**Comment by digama0** [comment]
bors d+

File: Mathlib/Tactic/Linarith/Preprocessing.lean (hunk 3 above)
Comment (semorrison) [reply Mathlib/Tactic/Linarith/Preprocessing.lean:332]: Surprising name, but okay. :-)

**Comment by semorrison** [comment]
bors merge

File: Mathlib/Tactic/Linarith/Frontend.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Tactic/Linarith/Frontend.lean]: This doesn't work: `whnf` is pretty violent:
```
0 < (t*(r + v) + s)*3*u
```
turns into `Rat.blt ... < true`...

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**